kernel/src/engine/arrow_utils.rs

```diff
@@ -195,7 +210,29 @@ pub(crate) fn ensure_data_types(
             let mut found_fields = 0;
             // ensure that for the fields that we found, the types match
             for (kernel_field, arrow_field) in mapped_fields.zip(arrow_fields) {
-                ensure_data_types(&kernel_field.data_type, arrow_field.data_type())?;
+                if check_nullability_and_metadata {
+                    if kernel_field.nullable != arrow_field.is_nullable() {
+                        return Err(Error::Generic(format!(
+                            "Field {} has nullablily {} in kernel and {} in arrow",
+                            kernel_field.name,
+                            kernel_field.nullable,
+                            arrow_field.is_nullable()
+                        )));
+                    }
+                    if &kernel_field.metadata_as_string() != arrow_field.metadata() {
```

Comment:
naive question: is there a way to do this more efficiently? saw that we are cloning metadata in this method

Reply:
Sadly both kernel and arrow metadata are just `HashMap` and we can't `impl PartialEq<HashMap<X,Y>> for HashMap<X,Z>` because `HashMap` isn't in our crate.

So I've done an impl to compare `MetadataValue` to a `String`, and written a helper method to compare the two hashmaps. Not as general but still pretty clean.